Debulleting Porch Architecture & Components - Engine (#1089)
nephio-project/porch: docs/content/en/docs/5_architecture_and_components/engine/functionality/validation-business-rules.md

- "unsupported lifecycle value: {value}" (for invalid values)

**Rationale:**

Packages must progress through the Draft/Proposed states before being published. This prevents bypassing the review/approval

> workflows and ensures all packages have a draft history.

Reviewer: one note to be aware of here is it also ensures all packages which pass to the published stage have a "successful" render of its kpt pipeline. so no failing render package can ever slip through.

Author: corrected

> workflows. This way all packages have a draft history.

Author: applied

- "invalid desired lifecycle value: {value}"

**Rationale:**

Published packages are immutable, meaning that their content cannot be changed. However, draft packages are mutable, since they are
work-in-progress. The lifecycle transitions must follow state machine rules (see

> [Lifecycle Management]({{% relref "/docs/5_architecture_and_components/engine/functionality/lifecycle-management.md" %}})).

Reviewer 2: I feel like this is a no-brainer:

> However, draft packages are mutable, since they are work-in-progress.

```suggestion
The work-in-progress draft packages are _mutable_, whilst the published ones are _immutable_, meaning that their content cannot be changed.
The lifecycle transitions must follow state machine rules (see [Lifecycle Management]({{% relref "/docs/5_architecture_and_components/engine/functionality/lifecycle-management.md" %}})).
```

Author: corrected

- "task list must not contain more than one task"

**Rationale:**

Only allowing one operation simplifies the creation workflow. However, you can add multiple tasks later with updates. The default init

> task provides sensible starting point.

Author: applied

> task provides s sensible starting point.

Reviewer: seems there is an extra "s" in the sentence here

Author: corrected

### Task Type Validation

| Valid task type | Description                               |
|-----------------|-------------------------------------------|
| init            | Create new package from scratch           |
| clone           | Copy package from upstream                |
| edit            | Create new revision from existing package |

> | upgrade         | Merge changes from new upstream version   |

Reviewer 2: Same suggestions that I made in the other file, we should deduplicate this in a separate PR

Author: corrected

- "package revision workspaceNames must be unique; package revision with name {name} in repo {repo} with workspaceName {workspace} already exists"

**Rationale:**

The workspace name is used to generate the Kubernetes object name, which must be unique within the namespace. This
prevents naming conflicts and confusion.

### Workspace Name Validation

> The workspace name must be a valid Kubernetes name. The repository and package is combined to create the object name.

Reviewer 2: They must be RFC 1123 label names I think https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names

Author: corrected

> Format: `{repo}-{path}-{package}-{workspace}`

Reviewer 2: This was incorrect even before, the separators are dots, not dashes. Unless this refers to something else...

```suggestion
Format: `{repo}.{path}.{package}.{workspace}`
```

Author: corrected

- "`clone` cannot create a new revision for package {package} that already exists in repo {repo}; make subsequent revisions using `copy`"

**Rationale:**

Cloning is for creating **new** packages from upstream, which is why existing packages should use `edit` or `copy` for new revisions.

> This prevents accidental overwriting of existing packages.

Reviewer 2: Should we point out that edit and copy are the same thing, just aliases?

Author: corrected

- "all source PackageRevisions of upgrade task must be published, {name} is not"

**Rationale:**

Upgrade performs a three-way merge (old upstream, new upstream, local). Source revisions must be stable (published) for reliable merge.

> This prevents upgrading from unstable draft versions

Author: applied

- Input format and structure

**Benefits:**

> It is fail fast (before expensive operations), gives clear error messages, and has no side effects on failure.

Reviewer 2: ```suggestion
The benefits of this are that it is fails fast (before expensive operations), gives clear error messages, and has no side effects on failure.
```

Author: corrected

- Package path overlaps

**Benefits:**

> It has access to the repository state, can check against existing data and has rollback mechanism available.

Reviewer 2: ```suggestion
The benefits of this are that it has access to the repository state, can check against existing data and has rollback mechanism available.
```

Author: corrected
